File: dp08_all_construct.py
def all_construct(target, word_bank):
    if target == "":
        return [[]]

    result = []

    for word in word_bank:
        if target.find(word) == 0:
            suffix = target[len(word):]
            suffixWays = all_construct(suffix, word_bank)
            targetWays = list(map(lambda w: [word,*w], suffixWays))
            result.extend(targetWays)

    return result


def main():
    print(all_construct('abcdef', ['ab', 'abc', 'cd', 'def', 'abcd', 'ef']))
    print(all_construct('', ['ab', 'abc', 'cd', 'def', 'abcd', 'ef']))
    print(all_construct('skateboard',  ['bo', 'rd', 'ate', 't', 'ska', 'sk', 'boar']))
    # print(all_construct('eeeeeeeeeeeeeeeeeeeeeeeeef',  ['e', 'ee', 'eee', 'eeee', 'eeeee', 'eeeeeee', 'eeeeeeeee']))

    # print(all_construct_dy('abcdef', ['ab', 'abc', 'cd', 'def', 'abcd'], {}))
    # print(all_construct_dy('skateboard',  ['bo', 'rd', 'ate', 't', 'ska', 'sk', 'boar'], {}))
    # print(all_construct_dy('eeeeeeeeeeeeeeeeeeeeeeeeeee',  ['e', 'ee', 'eee', 'eeee', 'eeeee', 'eeeeeee', 'eeeeeeee'], {}))

main()

File: dp08_allConstruct.go
package main

import (
	"strings"
)

func main() {
	println(allConstruct("abcdef", []string{"ab", "abc", "cd", "def", "abcd"}))
	println(allConstruct("skateboard", []string{"bo", "rd", "ate", "t", "ska", "sk", "boar"}))
	println(allConstruct("eeeeeeeeeeeeeeeeeeeeeeeeeeeef", []string{"e", "ee", "eee", "eeee", "eeeee", "eeeeeee", "eeeeeeee"}))
	// println(allConstructDy("abcdef", []string{"ab", "abc", "cd", "def", "abcd"}, map[string]int{}))
	// println(allConstructDy("skateboard", []string{"bo", "rd", "ate", "t", "ska", "sk", "boar"}, map[string]int{}))
	// println(allConstructDy("eeeeeeeeeeeeeeeeeeeeeeeeeeeef", []string{"e", "ee", "eee", "eeee", "eeeee", "eeeeeee", "eeeeeeee"}, map[string]int{}))
}

func allConstruct(target string, wordBank []string) [][]string {
	if target == "" {
		return [][]string{[]string{}}
	}

	var result [][]string

	for _, word := range wordBank {
		if strings.HasPrefix(target, word) {
			suffix := target[len(word):]
			suffixWays := allConstruct(suffix, wordBank)
			for _, ways := range suffixWays {
				result = append(result, append([]string{word}, ways...))
			}
		}
	}
	return result
}
